nifi-nar-bundles/nifi-cassandra-bundle/nifi-cassandra-processors/src/main/java/org/apache/nifi/processors/cassandra/PutCassandraQL.java

```diff
@@ -0,0 +1,403 @@
+                                    + "', which cannot be converted into the necessary data type: " + paramType, e);
+                        }
+                    }
+                }
+
+                try {
+                    ResultSetFuture future = connectionSession.executeAsync(boundStatement);
+                    if (statementTimeout > 0) {
+                        future.getUninterruptibly(statementTimeout, TimeUnit.MILLISECONDS);
+                    } else {
+                        future.getUninterruptibly();
+                    }
+                    // Emit a Provenance SEND event
+                    final long transmissionMillis = TimeUnit.NANOSECONDS.toMillis(System.nanoTime() - startNanos);
+
+                    // This isn't a real URI but since Cassandra is distributed we just use the cluster name
+                    String transitUri = "cassandra://" + connectionSession.getCluster().getMetadata().getClusterName();
+                    session.getProvenanceReporter().send(flowFile, transitUri, transmissionMillis, true);
+                    session.transfer(flowFile, REL_SUCCESS);
+
+                } catch (final TimeoutException e) {
+                    throw new ProcessException(e);
+                }
+
+
+            } catch (final NoHostAvailableException nhae) {
+                getLogger().error("No host in the Cassandra cluster can be contacted successfully to execute this statement", nhae);
+                // Log up to 10 error messages. Otherwise if a 1000-node cluster was specified but there was no connectivity,
+                // a thousand error messages would be logged. However we would like information from Cassandra itself, so
+                // cap the error limit at 10, format the messages, and don't include the stack trace (it is displayed by the
+                // logger message above).
+                getLogger().error(nhae.getCustomMessage(10, true, false));
+                flowFile = session.penalize(flowFile);
+                session.transfer(flowFile, REL_RETRY);
+
+            } catch (final QueryExecutionException qee) {
+                logger.error("Cannot execute the statement with the requested consistency level successfully", qee);
+                // The incoming flow file should be penalized
+                flowFile = session.penalize(flowFile);
+                session.transfer(flowFile, REL_RETRY);
+
+            } catch (final QueryValidationException qve) {
+                logger.error("The CQL statement {} is invalid due to syntax error, authorization issue, or another "
+                                + "validation problem; routing {} to failure",
+                        new Object[]{cql, flowFile}, qve);
+                flowFile = session.penalize(flowFile);
+                session.transfer(flowFile, REL_FAILURE);
+
+            } catch (final ProcessException e) {
+                if (flowFile == null) {
```

Comment:
Since PutCassandraQL is not a source processor the FlowFile will never be null here.